server/services/sequence_service_test.go

```diff
@@ -0,0 +1,223 @@
+	pb "github.com/googleapis/gapic-showcase/server/genproto"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func TestSequenceEmpty(t *testing.T) {
+	s := NewSequenceServer()
+
+	seq, err := s.CreateSequence(context.Background(), &pb.CreateSequenceRequest{})
+	if err != nil {
+		t.Errorf("CreateSequence(empty): unexpected err %+v", err)
+	}
+
+	timeout := 5 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	e, err := s.AttemptSequence(ctx, &pb.AttemptSequenceRequest{Name: seq.GetName()})
+	if err != nil {
+		t.Errorf("AttemptSequence(empty): unexpected err %+v", err)
+	}
+
+	if e == nil {
+		t.Errorf("AttemptSequence(empty): unexpected nil Empty response")
+	}
+
+	r := report(seq.GetName())
+	report, err := s.GetSequenceReport(context.Background(), &pb.GetSequenceReportRequest{Name: r})
+	if err != nil {
+		t.Errorf("GetSequenceReport(empty): unexpected err %+v", err)
+	}
+
+	attempts := report.GetAttempts()
+	if len(attempts) != 1 {
+		t.Errorf("%s: expected number of attempts to be 1 but was %d", t.Name(), len(attempts))
+	}
+
+	a := attempts[0]
+	ad := a.GetAttemptDeadline().AsTime()
+	d, _ := ctx.Deadline()
+
+	if !ad.Equal(d) {
+		t.Errorf("%s: server deadline = %v client deadline = %v", t.Name(), ad, d)
+	}
+}
+
+func TestSequenceRetry(t *testing.T) {
+	s := NewSequenceServer()
+	responses := []*pb.Sequence_Response{
+		{
+			Status: status.New(codes.Unavailable, "Unavailable").Proto(),
+			Delay:  ptypes.DurationProto(1 * time.Second),
+		},
+		{
+			Status: status.New(codes.Unavailable, "Unavailable").Proto(),
+			Delay:  ptypes.DurationProto(2 * time.Second),
+		},
+		{
+			Status: status.New(codes.OK, "OK").Proto(),
+		},
+	}
+
+	seq, err := s.CreateSequence(context.Background(), &pb.CreateSequenceRequest{
+		Sequence: &pb.Sequence{Responses: responses},
+	})
+	if err != nil {
+		t.Errorf("CreateSequence(retry): unexpected err %+v", err)
+	}
+
+	timeout := 5 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	delay := 100 * time.Millisecond
+	for n, r := range responses {
+		res := status.FromProto(r.GetStatus())
+		_, err = s.AttemptSequence(ctx, &pb.AttemptSequenceRequest{Name: seq.GetName()})
+		if c := status.Code(err); c != res.Code() {
+			t.Errorf("%s: status #%d was %v wanted %v", t.Name(), n, c, res.Code())
+		}
+
+		if n != len(responses)-1 {
+			time.Sleep(delay)
+			delay *= 2
+		}
+	}
+
+	r := report(seq.GetName())
+	report, err := s.GetSequenceReport(context.Background(), &pb.GetSequenceReportRequest{Name: r})
+	if err != nil {
+		t.Errorf("GetSequenceReport(retry): unexpected err %+v", err)
+	}
+
+	attempts := report.GetAttempts()
+	if len(attempts) != len(responses) {
+		t.Errorf("%s: expected number of attempts to be %d but was %d", t.Name(), len(responses), len(attempts))
+	}
+
+	d, _ := ctx.Deadline()
+	for n, a := range attempts {
+		if a.GetAttemptNumber() != int32(n) {
```

Comment:
nit: here and below, I suggest `if got, want := ...; got !=  want {` to avoid code duplication

Reply:
Good idea! Done.